std/string.cup

```diff
@@ -0,0 +1,268 @@
+    self.len = self.len + sl.len;
+    
+    for (let i = first; i < self.len; ++i){
+        self.chars[i] = sl::get(i-first);
+    }
+    
+    self.chars[self.len] = '\0';
+}
+
+method str::len(): i32{
+    return self.len;
+}
+
+method str::clone(): str*{
+    let clone: str* = new str(self.chars);
+    return clone;
+}
+
+method str::get(idx: i32): char{
+    if (idx < self.len){
+        return self.chars[idx];
+    } else {
+        die(here, "str::get: index out of bounds");
+    }
+}
+
+method str::slice(from: i32, to: i32): slice*{
+    if (to > self.len){
+        die(here, "str::slice: index out of bounds");
+    }
+    if (to < 0 || from < 0){
+        die(here, "str::slice: indexes cannot be negative nums");
+    }
+    
+    // contains index "to" too
+    let s: slice* = new slice(&self.chars[from], (to+1)-from, self); 
+    return s;
+    
+}
+
+method str::split(splitter: char): Vector*{
+    let start_slice = 0;
+    let slices: Vector* = vector_new();
+    for (let i = 0; i < self.len; ++i){
+        if (self::get(i) == splitter){
+            if (!(i == start_slice)){
+                let slice_: slice* = self::slice(start_slice, i-1);
+                slices::push(slice_);
+                start_slice = i + 1;
+            }
+        }
+    }
+    
+    if (start_slice < self.len){
+        let slice_: slice* = self::slice(start_slice, self.len-1);
+        slices::push(slice_);
+    }
+    
+    return slices;
+}
+
+method str::startswith_cstr(prefix: char*): i32{
+    return strstartswith(self.chars, prefix);
+}
+
+method str::startswith(prefix: str*): i32{
+    return self::startswith_cstr(prefix.chars);
+}
+
+method str::equals_cstr(other: char*): i32{
+    return streq(self.chars, other);
+}
+
+method str::equals(other: str*): i32{
+    return self::equals_cstr(other.chars);
+}
+
+// Checks whether self contains other and returns index of start of the other in self if it contains other.
+method str::contains_idx(other: str*): i32{
+    if (other.len == 0){
+        die(here, "str::contains: other has to have length greater than 0");
+    }
+    
+    for (let i = 0; i < self.len; ++i){
+        if (self::get(i) == other::get(0)){
+            let to = i+other.len-1;
+            if (to >= self.len){
+                break;
+            }
+            let slice_: slice* = self::slice(i, to);
+            
+            let contains = true;
+            for (let i_ = 0; i_ < slice_.len; ++i_){
+                if (slice_::get(i_) != other::get(i_)){
+                    contains = false;
+                    break;
+                }
+            }
+            
+            if (contains){
+                return i;
+            }
+            
+            free(slice_);
+        }
+    }
+    
+    return -1;
+}
+
+method str::contains(other: str*): i32{
+    let result = self::contains_idx(other);
+    
+    if (result == -1){
+        return false;
+    }
+    else 
+        return true;
+}
+
+method str::drop();
```

Comment:
Can just move the method definition up here since it doesn't need to use any of the other methods.

Reply:
Okay